File: app/store.py
"""
In-memory store for delivery records. Lets callers query what happened to
an event after dispatch.
 
Not persisted to disk, a restart clears history.
"""

from __future__ import annotations
import asyncio
from datetime import time

from app.models import EventRecord

class EventStore:
    def __init__(self, ttl_seconds: float) -> None:
        self._records: dict[str, tuple[float, EventRecord]] = {}
        self._lock = asyncio.Lock()
        self._ttl = ttl_seconds

    async def save(self, record: EventRecord) -> None:
        async with self._lock:
            expiry = time.monotonic() + self._ttl
            self._records[record.event_id] = (expiry, record)

    async def get(self, event_id: str) -> EventRecord | None:
        async with self._lock:
            entry = self._records.get(event_id)
            if entry is None: 
                return None # If no record, return None
            expiry, record = entry
            if time.monotonic() > expiry: # If expired, remove record and return None
                del self._records[event_id]
                return None
            return record
        
    async def all(self) -> list[EventRecord]:
        now = time.monotonic()
        async with self._lock:
            stale = [k for k, (exp, _) in self._records.items() if now > exp]
            for k in stale:
                del self._records[k]
            return [record for _, record in self._records.values()]
        
    async def evict_expired(self) -> int:
        """Called by the background task to remove expired records"""
        now = time.monotonic()
        async with self._lock:
            stale = [k for k, (exp, _) in self._records.items() if now > exp]
            for k in stale:
                del self._records[k]
            return len(stale)
        
    def __len__(self) -> int:
        return len(self._records)

File: internal/store/store.go
// In-memory store for webhooks
package store

import (
	"sync"
	"time"
	"webhook-dispatcher/internal/models"
)

type entry struct {
	record 		models.EventRecord
	expiresAt 	time.Time
}

type EventStore struct {
	mu 		sync.RWMutex
	records map[string]entry
	ttl 	time.Duration
}

func New(ttl time.Duration) *EventStore {
	return &EventStore{
		records: make(map[string]entry),
		ttl: ttl,
	}
}

func (s *EventStore) Save(record models.EventRecord) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.records[record.EventID] = entry{
		record:    record,
		expiresAt: time.Now().Add(s.ttl),
	}
}

func (s *EventStore) Get(id string) (models.EventRecord, bool) {
	s.mu.RLock()
	e, ok := s.records[id]
	s.mu.RUnlock()

	if !ok || time.Now().After(e.expiresAt) {
		return models.EventRecord{}, false
	}
	return e.record, true
}

func (s *EventStore) All() []models.EventRecord {
	now := time.Now()
	s.mu.RLock()
	defer s.mu.RUnlock()

	out := make([]models.EventRecord, 0, len(s.records))
	for _, e := range s.records {
		if !now.After(e.expiresAt) { // Prune expired records from the list
			out = append(out, e.record)
		}
	}
	return out
}

func (s *EventStore) EvictExpired() int {
	now := time.Now()
	s.mu.Lock()
	defer s.mu.Unlock()

	var evicted int
	for id, e := range s.records {
		if now.After(e.expiresAt) {
			delete(s.records, id)
			evicted++
		}
	}
	return evicted
}

func (s *EventStore) Len() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.records)
}
